File: epi_judge_python/directory_path_normalization.py
from test_framework import generic_test


def shortest_equivalent_path(path: str) -> str:
    if len(path) == 1 and path == '.':
        return path
    root = True if path and path[0] == '/' else False
    stack = []
    for x in path.split('/'):
        if x == '..':
            if len(stack) > 0 and stack[-1] != '..':
                stack.pop()
            else:
                stack.append(x)
        elif x and x != '.':
            stack.append(x)
    return ('/' if root else '') + '/'.join(stack)


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('directory_path_normalization.py',
                                       'directory_path_normalization.tsv',
                                       shortest_equivalent_path))

File: epi_judge_cpp/directory_path_normalization.cc
#include <sstream>
#include <string>

#include "test_framework/generic_test.h"
using std::string;
string ShortestEquivalentPath(const string &path) {
  auto ss = std::stringstream(path);
  std::string token;
  std::stack<string> lifo;
  bool root = path.front() == '/';

  while (getline(ss, token, '/')) {
    if (token == "..") {
      if (lifo.empty() || lifo.top() == "..") {
        lifo.push(token);
      } else {
        lifo.pop();
      }
    } else if (!token.empty() && token != ".") {
      lifo.push(token);
    }
  }

  string spath;
  while (!lifo.empty()) {
    spath = spath.empty() ? lifo.top() : lifo.top() + "/" + spath;
    lifo.pop();
  }
  if (root) {
    spath = "/" + spath;
  }
  return spath.empty() ? "/" : spath;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"path"};
  return GenericTestMain(args, "directory_path_normalization.cc",
                         "directory_path_normalization.tsv",
                         &ShortestEquivalentPath, DefaultComparator{},
                         param_names);
}
